packages/agents-core/src/runImplementation.ts

```diff
@@ -68,21 +81,78 @@ export function processModelResponse<TContext>(
   const runHandoffs: ToolRunHandoff[] = [];
   const runFunctions: ToolRunFunction<TContext>[] = [];
   const runComputerActions: ToolRunComputer[] = [];
+  const runMCPApprovalRequests: ToolRunMCPApprovalRequest[] = [];
   const toolsUsed: string[] = [];
   const handoffMap = new Map(handoffs.map((h) => [h.toolName, h]));
   const functionMap = new Map(
     tools.filter((t) => t.type === 'function').map((t) => [t.name, t]),
   );
   const computerTool = tools.find((t) => t.type === 'computer');
+  const mcpToolMap = new Map(
+    tools
+      .filter((t) => t.type === 'hosted_tool' && t.providerData?.type === 'mcp')
+      .map((t) => t as HostedMCPTool)
+      .map((t) => [t.providerData.serverLabel, t]),
+  );
 
   for (const output of modelResponse.output) {
     if (output.type === 'message') {
       if (output.role === 'assistant') {
         items.push(new RunMessageOutputItem(output, agent));
       }
     } else if (output.type === 'hosted_tool_call') {
-      items.push(new RunToolCallItem(output, agent));
-      toolsUsed.push(output.name);
+      if (
+        output.providerData?.type === 'mcp_approval_request' ||
+        output.name === 'mcp_approval_request'
+      ) {
+        // Hosted remote MCP server support
+        const providerData =
+          output.providerData as ProviderData.HostedMCPApprovalRequest;
+        const mcpServerLabel = providerData.serverLabel;
+        const mcpServerTool = mcpToolMap.get(mcpServerLabel);
+        if (mcpServerTool !== undefined) {
+          const toolName = JSON.stringify({
```

Comment:
Having both server label and tool name is helpful for MCP server usage, so I've included both in JSON format.